packages/wallets/src/components/ModalProvider/ModalProvider.tsx

```diff
@@ -32,17 +31,18 @@ export const useModal = () => {
 const ModalProvider = ({ children }: React.PropsWithChildren<unknown>) => {
     const modalRef = useRef<HTMLDivElement>(null);
     const [content, setContent] = useState<React.ReactNode | null>();
-    const modalState = useRef<TModalState>();
+    const [modalState, updateModalState] = useState<Map<keyof TModalState, TModalState[keyof TModalState]>>(new Map());
     const { isDesktop, isMobile } = useDevice();
 
     const rootRef = useRef<HTMLElement>(document.getElementById('wallets_modal_root'));
     const rootResponsiveRef = useRef<HTMLElement | null>(document.getElementById('wallets_modal_responsive_root'));
 
-    const setModalState = (newModalState: Partial<TModalState>) => {
-        modalState.current = {
-            ...modalState.current,
-            ...newModalState,
-        };
+    const getModalState = <T extends keyof TModalState>(key: T): TModalState[T] => {
+        return modalState.get(key) as TModalState[T];
+    };
+
+    const setModalState = <T extends keyof TModalState>(key: T, value: TModalState[T]) => {
+        updateModalState(new Map(modalState.set(key, value)));
```

Comment:
is there any reason for wrapping modalSate in extra map? looks a bit excessive

Reply:
We would need to treat the `modalState` as immutable, thus to update the current map we would need to create a new instance of it